ksql-api/src/main/java/io/confluent/ksql/api/server/BufferedPublisher.java

```diff
@@ -0,0 +1,284 @@
+
+  private static final Logger log = LoggerFactory.getLogger(BufferedPublisher.class);
+  public static final int SEND_MAX_BATCH_SIZE = 10;
+  public static final int DEFAULT_BUFFER_MAX_SIZE = 100;
+
+  private final Context ctx;
+  private final Queue<T> buffer = new ArrayDeque<>();
+  private final int bufferMaxSize;
+  private Subscriber<? super T> subscriber;
+  private long demand;
+  private boolean cancelled;
+  private Runnable drainHandler;
+  private boolean complete;
+  private boolean completing;
+
+  /**
+   * Construct a BufferedPublisher
+   *
+   * @param ctx The Vert.x context to use for the publisher - the publisher code must always be
+   *            executed on this context. This ensures the code is never executed concurrently by
+   *            more than one thread.
+   */
+  public BufferedPublisher(final Context ctx) {
+    this(ctx, Collections.emptySet(), DEFAULT_BUFFER_MAX_SIZE);
+  }
+
+  /**
+   * Construct a BufferedPublisher
+   *
+   * @param ctx           The Vert.x context to use for the publisher
+   * @param initialBuffer A collection of elements to initialise the buffer with
+   */
+  public BufferedPublisher(final Context ctx, final Collection<T> initialBuffer) {
+    this(ctx, initialBuffer, DEFAULT_BUFFER_MAX_SIZE);
+  }
+
+  /**
+   * Construct a BufferedPublisher
+   *
+   * @param ctx           The Vert.x context to use for the publisher
+   * @param bufferMaxSize Indicative max number of elements to store in the buffer. Note that this
+   *                      is not enforced, but it used to determine what to return from the accept
+   *                      method so the caller can stop sending more and set a drainHandler to be
+   *                      notified when the buffer is cleared
+   */
+  public BufferedPublisher(final Context ctx, final int bufferMaxSize) {
+    this(ctx, Collections.emptySet(), bufferMaxSize);
+  }
+
+  /**
+   * Construct a BufferedPublisher
+   *
+   * @param ctx           The Vert.x context to use for the publisher
+   * @param initialBuffer A collection of elements to initialise the buffer with
+   * @param bufferMaxSize Indicative max number of elements to buffer
+   */
+  public BufferedPublisher(final Context ctx, final Collection<T> initialBuffer,
+      final int bufferMaxSize) {
+    this.ctx = ctx;
+    this.buffer.addAll(initialBuffer);
+    this.bufferMaxSize = bufferMaxSize;
+  }
+
+  /**
+   * Provide an element to the publisher. The publisher will attempt to deliver it to it's
+   * subscriber (if any). The publisher will buffer it internally if it can't deliver it
+   * immediately.
+   *
+   * @param t The element
+   * @return true if the internal buffer is 'full'. I.e. if number of elements is >= bufferMaxSize.
+   */
+  public boolean accept(final T t) {
+    checkContext();
+    if (completing) {
+      throw new IllegalStateException("Cannot call accept after complete is called");
+    }
+    if (demand == 0 || cancelled) {
+      buffer.add(t);
```

Comment:
if it's cancelled, why would I add it to the buffer?

Reply:
Ack